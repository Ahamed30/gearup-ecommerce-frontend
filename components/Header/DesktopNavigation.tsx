import React from "react";
import Link from "next/link";
import { rubik } from "@/styles";

export const DesktopNavigation = () => {
  const fontRubik = rubik?.className;

  return (
    <div
      className={`${fontRubik} hidden font-semibold text-base w-full lg:block flex-shrink-0 flex flex-start items-center justify-start lg:w-auto`}
    >
      <Link
        href="#responsive-header"
        className="block mt-4 lg:inline-block lg:mt-0 text-black hover:text-black mr-4"
      >
        New DropsðŸ”¥
      </Link>
      <a
        href="#responsive-header"
        className="block mt-4 lg:inline-block lg:mt-0 text-black hover:text-black mr-4"
      >
        Apparels
      </a>
    </div>
  );
};
