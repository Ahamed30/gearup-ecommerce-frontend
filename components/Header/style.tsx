import { rubik } from "@/styles";

const fontRubik = rubik?.className;

export const navLinkContentClass = `${fontRubik} hidden lg:flex font-semibold text-base absolute left-0 top-3`;
