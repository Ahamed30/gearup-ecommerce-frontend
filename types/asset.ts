import { Key } from "react";

export interface AssetType {
  id?: Key;
  name?: String;
  url?: String;
}
