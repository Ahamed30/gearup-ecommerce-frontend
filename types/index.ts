export * from "./header";
export * from "./link";
export * from "./footer";
export * from "./linkColumn";
export * from "./subscribeForm";
export * from "./asset";
export * from "./homePage";
export * from "./productCard";
