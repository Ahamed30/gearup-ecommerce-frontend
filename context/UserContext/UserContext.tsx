import {
  createUserWithEmailAndPassword,
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
  updateProfile,
} from "firebase/auth";
import { useRouter } from "next/navigation";
import { createContext, useContext, useEffect, useState } from "react";
import type { ReactNode } from "react";
import { auth } from "@/firebase/firebaseConfig";
import { AddressType, UserType } from "@/types";
import { useApp } from "../AppContext";

interface UserContextType {
  user: UserType | null;
  handleChangeData: (e: {
    target: { name: string; value: string; type: string; checked: boolean };
  }) => void; // when updating via backend these can be update to Promise<void> from void
  updateAddress: (address: AddressType, typeOfAddress: string) => void;
  isLoggedIn: boolean;
  handleLogin: (email: string, password: string) => void;
  handleSignUp: (
    name: string,
    email: string,
    password: string
  ) => Promise<boolean>;
  handleSignOut: () => void;
  setIsLoggedIn: (loginStatus: boolean) => void;
}

//@ts-expect-error: ignore initial context creation
const UserContext = createContext<UserContextType>(null);

export const useUser = () => useContext(UserContext);

export interface UserContextProviderProps {
  children: ReactNode;
}

export const UserContextProvider = ({ children }: UserContextProviderProps) => {
  const [user, setUser] = useState<UserType>({
    email: "",
    deliveryAddress: null,
    phoneNumber: 1234567890,
    billingAddress: null,
    sameBillingDelivery: true,
    receiveNewsletter: false,
  });

  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const router = useRouter();
  const { setIsLoading } = useApp();

  useEffect(() => {
    onAuthStateChanged(auth, (user) => {
      if (user) {
        setUser((prevData) => {
          return { ...prevData, email: String(user.email) };
        });
        setIsLoggedIn(true);
      } else {
        setUser((prevData) => {
          return { ...prevData, email: "" };
        });
      }
    });
  }, []);

  const handleChangeData = (e: {
    target: { name: string; value: string; type: string; checked: boolean };
  }) => {
    const { name, value, type, checked } = e.target;
    setUser((prevData) => ({
      ...prevData,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const updateAddress = (address: AddressType, typeOfAddress: string) => {
    setUser((prevCart) => {
      const updatedUser: UserType = { ...prevCart };

      if (typeOfAddress === "deliveryAddress") {
        updatedUser.deliveryAddress = address;
      } else {
        updatedUser.billingAddress = address;
      }

      return updatedUser;
    });
  };

  const handleLogin = async (email: string, password: string) => {
    setIsLoading(true);
    await signInWithEmailAndPassword(auth, email, password)
      .then(async (userCredential) => {
        fetch("/api/login", {
          method: "POST",
          headers: {
            Authorization: `Bearer ${await userCredential.user.getIdToken()}`,
          },
        }).then((response) => {
          if (response.status == 200) {
            setIsLoggedIn(true);
            return response;
          }
        });
      })
      .catch((error) => console.error(error));
    window.location.replace(sessionStorage.getItem("previousPage") ?? "/");
    setIsLoading(false);
  };

  const handleSignUp = async (
    name: string,
    email: string,
    password: string
  ) => {
    setIsLoading(true);
    const userCredential = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    )
      .then(async (userCredential) => {
        handleLogin(email, password);
        return userCredential;
      })
      .catch((error) => {
        const errorCode = error.code;
        const errorMessage = error.message;
        if (
          errorCode === "auth/email-already-exists" ||
          errorCode === "auth/email-already-in-use"
        ) {
          return false;
        }
        console.error("Unable to register", errorMessage);
      });

    if (userCredential) {
      await updateProfile(userCredential.user, {
        displayName: name,
      });
    }
    router.push(sessionStorage.getItem("previousPage") ?? "/");
    setIsLoading(false);
    return true;
  };

  const handleSignOut = async () => {
    setIsLoading(true);
    await signOut(auth);

    await fetch("/api/logout", {
      method: "POST",
    }).catch((err) => {
      console.error("Failed to SignOut", err);
    });
    window.location.reload();
    setIsLoggedIn(false);
    setIsLoading(false);
  };

  return (
    <UserContext.Provider
      value={{
        user,
        handleChangeData,
        updateAddress,
        isLoggedIn,
        setIsLoggedIn,
        handleLogin,
        handleSignUp,
        handleSignOut,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};
