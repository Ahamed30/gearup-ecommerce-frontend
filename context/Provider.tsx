"use client";

import { SessionProvider } from "next-auth/react";
import type { ReactNode } from "react";
import { AppContextProvider } from "./AppContext";
import { CartContextProvider } from "./CartContext";
import { UserContextProvider } from "./UserContext";

interface ProviderProps {
  children: ReactNode;
}

export const Provider = ({ children }: ProviderProps) => {
  return (
    <SessionProvider>
      <AppContextProvider>
        <UserContextProvider>
          <CartContextProvider>{children}</CartContextProvider>
        </UserContextProvider>
      </AppContextProvider>
    </SessionProvider>
  );
};
