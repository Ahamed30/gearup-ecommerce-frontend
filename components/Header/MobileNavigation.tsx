import Link from "next/link";
import { signOut } from "next-auth/react";
import { useUser } from "@/context/UserContext";
import { rubik } from "@/styles";
import { LinkType } from "@/types";
import { UnStyledButton } from "../Button";
import { MobileSearchBox } from "../SearchBox";

interface MobileNavigationProps {
  navLinks?: LinkType[];
}

export const MobileNavigation = ({ navLinks }: MobileNavigationProps) => {
  const fontRubik = rubik?.className;
  const { isLoggedIn } = useUser();

  const containerClassName = `${fontRubik} w-full font-semibold text-base pt-4 lg:hidden block`;

  const navLinkContent = navLinks?.map((navLink, index) => {
    const { link, name } = navLink;
    // {To check data issue for the above id}
    return (
      <Link
        className="block mt-1 text-black hover:text-[#8F8C91] hover:bg-gray-100 px-1 py-3"
        href={`${link}`}
        key={index}
      >
        {name}
      </Link>
    );
  });

  return (
    <div className={containerClassName} role="mobileSearch">
      <MobileSearchBox />
      {navLinkContent}
      {isLoggedIn ? (
        <UnStyledButton
          className="w-full text-base underline underline-offset-4 px-1 pt-3 border-t-2 border-gray-400 text-left"
          onClick={() => signOut()}
        >
          Sign Out
        </UnStyledButton>
      ) : null}
    </div>
  );
};
