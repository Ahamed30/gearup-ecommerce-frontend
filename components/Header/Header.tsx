"use client";

import Image from "next/image";
import { useRouter } from "next/navigation";
import { signOut } from "next-auth/react";
import { useState } from "react";
import { useCart } from "@/context/CartContext";
import { useUser } from "@/context/UserContext";
import { graduate } from "@/styles";
import { HeaderType } from "@/types";
import { DesktopNavigation } from "./DesktopNavigation";
import { MobileNavigation } from "./MobileNavigation";
import { UnStyledButton } from "../Button";
import { DesktopSearchBox } from "../SearchBox";

interface HeaderProps {
  data: HeaderType;
}

export const Header = ({ data }: HeaderProps) => {
  const [isMobileNavBarOpen, setIsMobileNavBarOpen] = useState(false);
  const [isDesktopSearchBoxOpen, setIsDesktopSearchBoxOpen] = useState(false);
  const { cart } = useCart();
  const router = useRouter();
  const { isLoggedIn } = useUser();
  const { navLinks } = data;

  const onClickMobileMenuBar: () => void = () => {
    setIsMobileNavBarOpen(!isMobileNavBarOpen);
  };

  const onClickMobileSearchBox = () => {
    setIsDesktopSearchBoxOpen(!isDesktopSearchBoxOpen);
  };

  return (
    <header className="flex flex-wrap border rounded-xl bg-[#FAFAFA] p-4 lg:p-6 mb-6 lg:mb-8">
      <div className="w-full flex justify-center content-center relative">
        <DesktopNavigation navLinks={navLinks} />
        <div
          className="block lg:hidden absolute left-0 top-4"
          onClick={onClickMobileMenuBar}
        >
          <Image alt="Nav Menu" height={20} src={"/nav-menu.svg"} width={20} />
        </div>
        <div className="flex justify-center items-center text-black">
          <span
            className={`${graduate?.className} font-bold text-[32px] cursor-pointer`}
            onClick={() => router.push("/")}
          >
            GearUp
          </span>
        </div>
        <div className="flex absolute right-0 top-3">
          <UnStyledButton
            className="hidden lg:block mr-[40px]"
            onClick={onClickMobileSearchBox}
          >
            <Image
              alt="navSearchIcon"
              height={20}
              src={"/search.svg"}
              width={20}
            />
          </UnStyledButton>
          <UnStyledButton
            className="mr-[16px] lg:mr-[40px]"
            onClick={() =>
              isLoggedIn ? router.push("/") : router.push("/auth/login")
            }
          >
            <Image alt="User Icon" height={20} src={"/user.svg"} width={20} />
          </UnStyledButton>
          {/* TODO: Need to change this to cart icon */}
          <UnStyledButton
            className="flex flex-col justify-center items-center px-[6px] lg:py-[4px] lg:px-[8px] border rounded-3xl bg-[#FFA52F]"
            id="cart"
            onClick={() => router.push("/cart")}
          >
            {cart?.itemCount ?? 0}
          </UnStyledButton>
        </div>
      </div>
      {isMobileNavBarOpen && <MobileNavigation navLinks={navLinks} />}
      <DesktopSearchBox isDesktopSearchBoxOpen={isDesktopSearchBoxOpen} />
      <UnStyledButton onClick={() => signOut()}>SignOut</UnStyledButton>
    </header>
  );
};
