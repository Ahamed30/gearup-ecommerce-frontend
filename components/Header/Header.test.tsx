// Header.test.tsx
import { fireEvent, getByRole, render, screen } from "@testing-library/react";
import { Header } from "./Header";
import { UserContextProvider } from "@/context/UserContext";
import { HeaderType } from "@/types";

// Mock the useRouter push function
const mockRouterPush = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

// Mock the useCart hook
jest.mock("@/context/CartContext", () => ({
  useCart: () => ({
    cart: {
      itemCount: 0,
    },
  }),
}));

// Test data for the Header component
const testData = {
  navLinks: [
    { name: "Home", link: "/" },
    { name: "About", link: "/about" },
    { name: "Contact", link: "/contact" },
  ],
};

const navLinks = [
  { id: "1", link: "/", name: "Home" },
  { id: "2", link: "/shop", name: "Shop" },
  { id: "3", link: "/about", name: "About" },
];

const renderComponent = (data: HeaderType = testData) => {
  return render(
    <UserContextProvider>
      <Header data={data} />
    </UserContextProvider>
  );
};

describe("Header", () => {
  it("should render without crashing", () => {
    const { container } = renderComponent(testData);
    expect(container).toMatchSnapshot();
  });

  it("should navigate to the home page when clicking on the GearUp logo", () => {
    renderComponent(testData);
    const logo = screen.getByText("GearUp");
    fireEvent.click(logo);
    expect(mockRouterPush).toHaveBeenCalledWith("/");
  });

  it("should display 0 when there are no items in cart", () => {
    const container = renderComponent({ navLinks: [] });
    const cartButton = container.getByRole("button", { name: "0" });
    expect(cartButton).toHaveTextContent("0");
    fireEvent.click(cartButton);
    expect(mockRouterPush).toHaveBeenCalledWith("/cart");
  });

  it("should open and close mobile navigation bar when clicked", () => {
    const { getByAltText, getByRole } = renderComponent(testData);
    const menuIcon = getByAltText("Nav Menu");
    fireEvent.click(menuIcon);
    const link1 = getByRole("mobileSearch");
    expect(link1).toBeInTheDocument();
    fireEvent.click(menuIcon);
    expect(link1).not.toBeInTheDocument();
  });

  it("should render desktop navigation", () => {
    const { getByText } = renderComponent({ navLinks });
    navLinks.forEach((navLink) => {
      const linkElement = getByText(navLink.name);
      expect(linkElement).toBeInTheDocument();
    });
  });
});
