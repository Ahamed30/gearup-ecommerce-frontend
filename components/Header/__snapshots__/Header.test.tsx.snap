// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render without crashing 1`] = `
<div>
  <header
    class="flex flex-wrap border rounded-xl bg-[#FAFAFA] p-4 lg:p-6 mb-6 lg:mb-8"
  >
    <div
      class="w-full flex justify-center content-center relative"
    >
      <div
        class="className hidden lg:flex font-semibold text-base absolute left-0 top-3"
      >
        <a
          class="block mt-4 lg:inline-block lg:mt-0 text-black hover:text-[#8F8C91] mr-[40px]"
          href="/"
        >
          Home
        </a>
        <a
          class="block mt-4 lg:inline-block lg:mt-0 text-black hover:text-[#8F8C91] mr-[40px]"
          href="/about"
        >
          About
        </a>
        <a
          class="block mt-4 lg:inline-block lg:mt-0 text-black hover:text-[#8F8C91] mr-[40px]"
          href="/contact"
        >
          Contact
        </a>
      </div>
      <div
        class="block lg:hidden absolute left-0 top-4"
      >
        <img
          alt="Nav Menu"
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/nav-menu.svg"
          style="color: transparent;"
          width="20"
        />
      </div>
      <div
        class="flex justify-center items-center text-black"
      >
        <span
          class="className font-bold text-[32px] cursor-pointer"
        >
          GearUp
        </span>
      </div>
      <div
        class="flex absolute right-0 top-3"
      >
        <button
          class="hidden lg:block mr-[40px]"
        >
          <img
            alt="navSearchIcon"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            src="/search.svg"
            style="color: transparent;"
            width="20"
          />
        </button>
        <button
          class="mr-[16px] lg:mr-[40px]"
        >
          <img
            alt="User Icon"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            src="/user.svg"
            style="color: transparent;"
            width="20"
          />
        </button>
        <button
          class="flex flex-col justify-center items-center px-[6px] lg:py-[4px] lg:px-[8px] border rounded-3xl bg-[#FFA52F]"
          id="cart"
        >
          0
        </button>
      </div>
    </div>
    <div
      class="items-center justify-center w-full mt-8 hidden hidden"
    >
      <div
        class="flex flex-col w-1/2"
      >
        <div
          class="px-2 py-2.5 rounded-lg border justify-start items-center inline-flex border-[#232321]"
        >
          <div
            class="flex items-center cursor-pointer"
          >
            <img
              alt="Search Icon"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/search.svg"
              style="color: transparent;"
              width="20"
            />
          </div>
          <input
            class="className w-full px-2 placeholder-[#79767C] text-[16px] font-normal tracking-wide bg-transparent focus:outline-none"
            placeholder="Search"
            style="color: rgb(31, 26, 36);"
            type="text"
          />
        </div>
      </div>
    </div>
    <button>
      SignOut
    </button>
  </header>
</div>
`;
